core/src/mast/serialization/mod.rs

```diff
@@ -166,34 +172,33 @@ impl Serializable for MastForest {
 
 impl Deserializable for MastForest {
     fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
-        let magic: [u8; 5] = source.read_array()?;
-        if magic != *MAGIC {
-            return Err(DeserializationError::InvalidValue(format!(
-                "Invalid magic bytes. Expected '{:?}', got '{:?}'",
-                *MAGIC, magic
-            )));
-        }
-
-        let version: [u8; 3] = source.read_array()?;
-        if version != VERSION {
-            return Err(DeserializationError::InvalidValue(format!(
-                "Unsupported version. Got '{version:?}', but only '{VERSION:?}' is supported",
-            )));
-        }
+        read_and_validate_magic(source)?;
+        read_and_validate_version(source)?;
 
-        let decorator_count = source.read_usize()?;
+        // Reading sections metadata
         let node_count = source.read_usize()?;
+        let decorator_count = source.read_usize()?;
+
+        // Reading procedure roots
         let roots: Vec<u32> = Deserializable::read_from(source)?;
+
+        // Reading nodes
+        let basic_block_data: Vec<u8> = Deserializable::read_from(source)?;
+        let mast_node_infos: Vec<MastNodeInfo> = node_infos_iter(source, node_count)
+            .collect::<Result<Vec<MastNodeInfo>, DeserializationError>>()?;
+
+        // Reading Decorators
         let decorator_data: Vec<u8> = Deserializable::read_from(source)?;
-        let node_data: Vec<u8> = Deserializable::read_from(source)?;
         let string_table: StringTable = Deserializable::read_from(source)?;
+        let decorator_infos: Vec<DecoratorInfo> =
+            decorator_infos_iter(source, decorator_count)
+                .collect::<Result<Vec<DecoratorInfo>, DeserializationError>>()?;
```

Comment:
```suggestion
        let decorator_infos = decorator_infos_iter(source, decorator_count);
```

Nit: Now that we have an iterator we don't need to `collect` it so we can avoid the extra allocation.

In the loop we can just do

```rust
for decorator_info in decorator_infos {
    let decorator_info = decorator_info?;
    // ...
```